### PR TITLE
nixos/oauth2_proxy_nginx: re-do virtualHosts option

### PR DESCRIPTION
**Pull request by Kranzes**
###### Description of changes

It is now possible to list all the webroot (per vhost) locations you want to be protected behind the OAuth2 proxy.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Kranzes** [comment]
I am wondering what changes are needed to make this backport-able to stable channel?

File: nixos/modules/services/security/oauth2_proxy_nginx.nix [hunk 1]
```diff
@@ -4,6 +4,10 @@ let
   cfg = config.services.oauth2_proxy.nginx;
 in
 {
+  imports = [
+    (mkRemovedOptionModule [ "services" "oauth2_proxy" "nginx" "virtualHosts" ] "This option has been removed in favor of 'services.oauth2_proxy.nginx.virtualHostLocations' which allows setting the proxy protection at a per-location level.")
```
Comment (SuperSandro2000) [inline nixos/modules/services/security/oauth2_proxy_nginx.nix:8]:
```suggestion
    (mkRemovedOptionModule [ "services" "oauth2_proxy" "nginx" "virtualHosts" ]
      "This option has been removed in favor of 'services.oauth2_proxy.nginx.virtualHostLocations' which allows setting the proxy protection at a per-location level.")
```

File: nixos/modules/services/security/oauth2_proxy_nginx.nix [hunk 2]
```diff
@@ -13,54 +17,66 @@ in
...
-          proxy_set_header Content-Length   "";
-          proxy_pass_request_body           off;
-        '';
-      };
-      locations."/".extraConfig = ''
-        auth_request /oauth2/auth;
-        error_page 401 = /oauth2/sign_in;
+  config = {
+    assertions = [{
+      assertion = !builtins.any (loc: builtins.elem loc (flatten (builtins.attrValues cfg.virtualHostLocations))) [ "/oauth2/" "/oauth2/auth" ];
```
Comment (SuperSandro2000) [inline nixos/modules/services/security/oauth2_proxy_nginx.nix:33]:
```suggestion
      assertion = !lib.any (loc: lib.elem loc (flatten (lib.attrValues cfg.virtualHostLocations))) [ "/oauth2/" "/oauth2/auth" ];
```
if there is a lib counterpart of a lib function that one should be used.

File: nixos/modules/services/security/oauth2_proxy_nginx.nix [hunk 3]
```diff
@@ -13,54 +17,66 @@ in
...
 
-        # pass information via X-User and X-Email headers to backend,
-        # requires running with --set-xauthrequest flag
-        auth_request_set $user   $upstream_http_x_auth_request_user;
-        auth_request_set $email  $upstream_http_x_auth_request_email;
-        proxy_set_header X-User  $user;
-        proxy_set_header X-Email $email;
+    services.nginx = mkIf (cfg.virtualHostLocations != { }) {
+      recommendedProxySettings = true; # needed because duplicate headers
+      virtualHosts = builtins.mapAttrs
```
Comment (SuperSandro2000) [inline nixos/modules/services/security/oauth2_proxy_nginx.nix:39]:
```suggestion
      virtualHosts = lib.mapAttrs
```

File: nixos/modules/services/security/oauth2_proxy_nginx.nix [hunk 4]
```diff
@@ -4,6 +4,11 @@ let
   cfg = config.services.oauth2_proxy.nginx;
 in
 {
+  imports = [
+    (mkRemovedOptionModule [ "services" "oauth2_proxy" "nginx" "virtualHosts" ]
+      "This option has been removed in favor of 'services.oauth2_proxy.nginx.virtualHostLocations' which allows setting the proxy protection at a per-location level.")
```
Comment (SuperSandro2000) [inline nixos/modules/services/security/oauth2_proxy_nginx.nix:9]: Can't we map this onto the new option?

Comment (Kranzes) [reply nixos/modules/services/security/oauth2_proxy_nginx.nix:9]: What do you mean?

Comment (SuperSandro2000) [reply nixos/modules/services/security/oauth2_proxy_nginx.nix:9]: Can we use mkRenamedOption instead?

**Comment by SuperSandro2000** [comment]
I did a proper solution to protect all locations in #275541 which depends on #273234. I am going to close this PR as discussed in https://github.com/NixOS/nixpkgs/pull/255023#issuecomment-1863595697 and because I think the idea proposed here is to brittle to do mistakes and requires to many manual extra settings to work properly.

**Pull request closed without merging**